File: tool/calc_js/verify.mjs
// Runs every calc_api.dart function through the COMPILED JavaScript and prints
// the results in exactly the shape tool/calc_js/verify.dart prints them.
//
// The bundle it loads is the committed one the site imports:
//   site/lib/generated/radar-calc.js
//
//   dart compile js -O4 -o <out>.js tool/calc_js/radar_calc.dart
//   dart run tool/calc_js/verify.dart            > vm.json
//   node tool/calc_js/verify.mjs <out>.js        > js.json
//   diff vm.json js.json
//
// An empty diff is the claim: one source of arithmetic, two runtimes.
import { readFileSync } from 'node:fs';

const bundle = process.argv[2];
if (!bundle) {
  console.error('usage: node tool/calc_js/verify.mjs <compiled bundle.js>');
  process.exit(2);
}

// The bundle is an ES MODULE, not a classic script: gen-calc-js.mjs appends an
// `export const radarCalc` footer so the site's bundler follows the file and
// importers get a value instead of reaching for a global. That footer is why
// evaluating it with vm.runInThisContext — which parses classic-script
// syntax — died on `Unexpected token 'export'`, and left this half of the
// guard unrun.
//
// It is imported as a data: URL rather than by path because module-ness in Node
// comes from the package scope, and site/package.json has no `"type": "module"`
// — so `import('.../radar-calc.js')` is only accepted by Node's module-syntax
// DETECTION fallback, which warns on stderr and is not something a guard should
// depend on. A data: URL is unambiguously a module, needs no temp file, and
// leaves the committed bundle byte-identical — which matters, because
// gen-calc-js.mjs --check gates the build on that file's hash.
//
// The EXPORT is what gets read, not globalThis: that is the value the site
// actually imports, so the footer's contract is verified here too.
const source = readFileSync(bundle, 'utf8');
const module = await import(
  'data:text/javascript;base64,' +
    Buffer.from(source, 'utf8').toString('base64')
);

const calc = module.radarCalc ?? globalThis.radarCalc;
if (!calc) {
  console.error(
    `${bundle} exported no radarCalc and set no global — did main() run?`
  );
  process.exit(1);
}

const f = JSON.parse(
  readFileSync(process.argv[3] ?? 'tool/calc_js/fixture.json', 'utf8')
);

const pick = (keys) =>
  JSON.stringify(Object.fromEntries(keys.map((k) => [k, f[k] ?? null])));

const out = {
  analytics: JSON.parse(calc.analytics(pick(['trades', 'capital', 'maxRiskPercent']))),
  stats: JSON.parse(calc.stats(pick(['trades', 'capital', 'maxRiskPercent']))),
  decisions: JSON.parse(
    calc.decisions(
      pick(['trades', 'capital', 'maxRiskPercent', 'today', 'waitingThresholdDays'])
    )
  ),
  scenarios: JSON.parse(
    calc.scenarios(
      pick(['trades', 'defaultTakeProfitPercent', 'defaultStopLossPercent'])
    )
  ),
  riskScore: JSON.parse(calc.riskScore(pick(['trade', 'capital', 'maxRiskPercent']))),
  tradeMetrics: JSON.parse(
    calc.tradeMetrics(pick(['trade', 'capital', 'maxRiskPercent']))
  ),
  sizing: JSON.parse(
    calc.sizing(pick(['capital', 'maxRiskPercent', 'entry', 'stop', 'userQty', 'budget']))
  ),
  smartTrade: JSON.parse(
    calc.smartTrade(
      pick([
        'capital',
        'maxRiskPercent',
        'takeProfitPercent',
        'stopLossPercent',
        'entryPrice',
        'userQty',
        'stopPrice',
        'targetPrice',
        'budget',
      ])
    )
  ),
  goalPlan: JSON.parse(
    calc.goalPlan(
      pick([
        'mode',
        'targetAmount',
        'monthlyDeposit',
        'years',
        'annualReturnPercent',
        'initialAmount',
      ])
    )
  ),
  projection: JSON.parse(
    calc.projection(pick(['trades', 'capital', 'targetAmount', 'maxRiskPercent']))
  ),
  entitlement: JSON.parse(
    calc.entitlement(pick(['plan', 'trialStartedAt', 'proUntil', 'now']))
  ),
  riskMath: JSON.parse(
    calc.riskMath(
      pick([
        'capital',
        'maxRiskPercent',
        'entry',
        'stop',
        'price',
        'a',
        'b',
        'ratio',
        'threshold',
        'riskPct',
      ])
    )
  ),
  checklist: JSON.parse(calc.checklist(JSON.stringify(f.checklistIds))),
};

process.stdout.write(JSON.stringify(out, null, 2));

File: tool/calc_js/verify.dart
// Runs every calc_api.dart function on the Dart VM and prints the results as
// one JSON document.
//
// tool/calc_js/verify.mjs does the same thing through the COMPILED JavaScript.
// Diff the two outputs and you have proof that the app's arithmetic and the
// site's arithmetic are the same arithmetic — not two implementations that
// currently agree.
import 'dart:convert';
import 'dart:io';

import 'calc_api.dart';

void main(List<String> args) {
  final path = args.isEmpty ? 'tool/calc_js/fixture.json' : args.first;
  final input = File(path).readAsStringSync();
  final f = jsonDecode(input) as Map<String, dynamic>;

  String pick(List<String> keys) =>
      jsonEncode({for (final k in keys) k: f[k]});

  final out = <String, Object?>{
    'analytics': jsonDecode(
      analytics(pick(['trades', 'capital', 'maxRiskPercent'])),
    ),
    'stats': jsonDecode(stats(pick(['trades', 'capital', 'maxRiskPercent']))),
    'decisions': jsonDecode(
      decisions(pick([
        'trades',
        'capital',
        'maxRiskPercent',
        'today',
        'waitingThresholdDays',
      ])),
    ),
    'scenarios': jsonDecode(
      scenarios(pick([
        'trades',
        'defaultTakeProfitPercent',
        'defaultStopLossPercent',
      ])),
    ),
    'riskScore': jsonDecode(
      riskScore(pick(['trade', 'capital', 'maxRiskPercent'])),
    ),
    'tradeMetrics': jsonDecode(
      tradeMetrics(pick(['trade', 'capital', 'maxRiskPercent'])),
    ),
    'sizing': jsonDecode(
      sizing(pick([
        'capital',
        'maxRiskPercent',
        'entry',
        'stop',
        'userQty',
        'budget',
      ])),
    ),
    'smartTrade': jsonDecode(
      smartTrade(pick([
        'capital',
        'maxRiskPercent',
        'takeProfitPercent',
        'stopLossPercent',
        'entryPrice',
        'userQty',
        'stopPrice',
        'targetPrice',
        'budget',
      ])),
    ),
    'goalPlan': jsonDecode(
      goalPlan(pick([
        'mode',
        'targetAmount',
        'monthlyDeposit',
        'years',
        'annualReturnPercent',
        'initialAmount',
      ])),
    ),
    'projection': jsonDecode(
      projection(pick(['trades', 'capital', 'targetAmount', 'maxRiskPercent'])),
    ),
    'entitlement': jsonDecode(
      entitlement(pick(['plan', 'trialStartedAt', 'proUntil', 'now'])),
    ),
    'riskMath': jsonDecode(
      riskMath(pick([
        'capital',
        'maxRiskPercent',
        'entry',
        'stop',
        'price',
        'a',
        'b',
        'ratio',
        'threshold',
        'riskPct',
      ])),
    ),
    'checklist': jsonDecode(checklist(jsonEncode(f['checklistIds']))),
  };

  // Sorted keys, two-space indent — the JS side prints the same shape, so a
  // plain byte diff is the whole test.
  stdout.write(const JsonEncoder.withIndent('  ').convert(webNumbers(out)));
}

/// Prints numbers the way the compiled bundle prints them.
///
/// Dart on the VM has `int` and `double`; dart2js has neither — every Dart
/// number becomes one JavaScript number. So `jsonEncode(2500.0)` writes
/// `2500.0` here and `2500` through the bundle. That is a disagreement about
/// how a runtime SPELLS a number, not about the number: `2500.0 == 2500`, and
/// the JS side has no way to spell it the long way even when it wants to.
///
/// Left alone it puts 35 lines of noise into a diff whose entire contract is
/// that it comes back empty — a guard that is always red is a guard nobody
/// reads. So integral doubles are narrowed to `int`, which is the spelling
/// both runtimes can produce.
///
/// Two deliberate exclusions, both so a real disagreement stays visible:
///
///   * **Negative zero.** `-0.0` and `0` are different values — that is why
///     the migration in CLAUDE.md §24 compared with `Object.is` — so it is
///     never folded into `0`. If one ever appears the diff goes red, which is
///     the correct outcome for a number-shape difference that is real.
///   * **Anything past 2^53**, where a JS number stops being an exact integer
///     and both runtimes start choosing between exponent and long form on
///     their own. Narrowing there would invent agreement instead of testing
///     for it.
Object? webNumbers(Object? value) {
  if (value is double &&
      value.isFinite &&
      value == value.roundToDouble() &&
      value.abs() <= 9007199254740992.0 &&
      !(value == 0 && value.isNegative)) {
    return value.toInt();
  }
  if (value is Map) {
    return {
      for (final entry in value.entries) entry.key: webNumbers(entry.value),
    };
  }
  if (value is List) {
    return value.map(webNumbers).toList();
  }
  return value;
}
